### PR TITLE
Add some notes on javadoc to CONTRIBUTING

### PR DESCRIPTION
**Pull request by nik9000**
I've been writing a bunch of javadoc lately and have some opinions that
I feel might be useful to share in CONTRIBUTING.md. I'm no expert
though.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -436,11 +436,16 @@ is to be helpful, not to turn writing code into a chore.
        regular comments in the code. Remember as well that Elasticsearch
        has extensive [user documentation](./docs), and it is not the role
        of Javadoc to replace that.
+        1. If a method's performance is "unexpected" then its good to call that
```
Comment (pugnascotia) [inline CONTRIBUTING.md:439]:
```suggestion
        * If a method's performance is "unexpected" then it's good to call that
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -436,11 +436,16 @@ is to be helpful, not to turn writing code into a chore.
...
+           out in the javadoc. Especially if its usually fast but sometimes
```
Comment (pugnascotia) [inline CONTRIBUTING.md:440]:
```suggestion
           out in the Javadoc. This is especially helpful if the method is usually fast but sometimes
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -436,11 +436,16 @@ is to be helpful, not to turn writing code into a chore.
...
+           very slow (shakes fist at caching).
    12. Please still try to make class, method or variable names as
        descriptive and concise as possible, as opposed to relying solely on
        Javadoc to describe something.
-   13. Use `@link` and `@see` to add references, either to related
-       resources in the codebase or to relevant external resources.
+   13. Use `@link` to add references to related resources in the codebase. Or
+       outside the code base.
+       1. `@see` is much more limited than `@link`. You can use it but most of
+          the time `@link` flows better.
```
Comment (pugnascotia) [inline CONTRIBUTING.md:448]: There's a good discussion here - https://stackoverflow.com/questions/10097199/javadoc-see-or-link - worth linking to it?

**Review by pugnascotia — APPROVED** [review]

File: CONTRIBUTING.md (hunk 3 above)
Comment (nik9000) [reply CONTRIBUTING.md:448]: Yeah - we can. I come down on the other side from this answer though. I think if its important you should mention it in the prose. If you mention something twice you should link the first time and not again. Either `@linkplain` or don't link at all.

**Comment by nik9000** [comment]
Backported with 15841b0.

**Pull request merged**